### PR TITLE
Ensure failures in lifecycle methods take precedence over earlier TestAbortedExceptions

### LINKED ISSUE #1313: Failures in lifecycle methods should take precedence over TestAbortedException
## Overview

In the following example, `abortedTest` will be reported as _aborted_, while `successfulTest` will be reported as _failed_ due to the exception in `afterEach()`.

Expected behavior: `abortedTest` should be reported as _failed_ with a suppressed `TestAbortedException`.

```java
class TestCase {

	@Test
	void abortedTest() {
		assumeFalse(true, "abortedTest");
	}

	@Test
	void successfulTest() {
	}

	@AfterEach
	void afterEach() {
		throw new RuntimeException("afterEach");
	}
}
```

## Deliverables

- [ ] Let other exceptions take precedence over `TestAbortedException`.

### PR DESCRIPTION
**Pull request by marcphilipp**
Fixes #1313.

---

### Definition of Done

- [X] There are no TODOs left in the code
- [X] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [X] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [X] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [X] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [X] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/SingleTestExecutor.java [hunk 1]
```diff
@@ -1,78 +0,0 @@
...
-import org.opentest4j.TestAbortedException;
-
-/**
- * {@code SingleTestExecutor} encapsulates the execution of a single test
- * wrapped in an {@link Executable}.
- *
- * @since 1.0
- * @see #executeSafely(Executable)
- */
-@API(status = MAINTAINED, since = "1.0")
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/SingleTestExecutor.java]: Just realized this has status _maintained_... In that case, I'll resurrect and deprecate it.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ThrowableCollector.java [hunk 2]
```diff
@@ -8,22 +8,27 @@
...
+import static org.junit.platform.engine.TestExecutionResult.aborted;
+import static org.junit.platform.engine.TestExecutionResult.failed;
+import static org.junit.platform.engine.TestExecutionResult.successful;
 
 import org.apiguardian.api.API;
-import org.junit.jupiter.api.function.Executable;
 import org.junit.platform.commons.util.ExceptionUtils;
 import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.engine.TestExecutionResult;
+import org.opentest4j.TestAbortedException;
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ThrowableCollector.java]: This has the consequence of introducing a mandatory dependency on the OTA for any implementation of `HierarchicalTestExecutor` even if such a `TestEngine` does not use or even support exceptions from the OTA.

Thus, I'm not so sure we would want to do that.

Perhaps an alternative would be to allow a caller of the relocated `ThrowableCollector` to specify what the caller considers to be an "overridable exception".

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ThrowableCollector.java [hunk 3]
```diff
@@ -57,7 +67,11 @@ private void add(Throwable t) {
 		if (this.throwable == null) {
 			this.throwable = t;
 		}
-		else {
+		else if (this.throwable instanceof TestAbortedException && !(t instanceof TestAbortedException)) {
+			t.addSuppressed(this.throwable);
+			this.throwable = t;
+		}
+		else if (throwable != t) {
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ThrowableCollector.java]: What's the use case for this last `else if` block?

I'm not seeing it in my browser. 😉

**Comment by marcphilipp** [comment]
@sbrannen I addressed your concerns in b68ee5a660788fc9525f6c558cd6966b9a5fffe5. PTAL! 🙂

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/SingleTestExecutor.java [hunk 4]
```diff
@@ -26,8 +26,11 @@
  *
  * @since 1.0
  * @see #executeSafely(Executable)
+ * @deprecated Please use {@link ThrowableCollector#execute} and
+ * {@link ThrowableCollector#toTestExecutionResult} instead.
  */
-@API(status = MAINTAINED, since = "1.0")
+@Deprecated
+@API(status = DEPRECATED, since = "1.2")
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/SingleTestExecutor.java]: This should be 1.3.

**Comment by sbrannen** [comment]
> I addressed your concerns in b68ee5a. PTAL! 🙂

Looks better now. Thanks.

**Pull request merged**